### PR TITLE
lockfile: add RecursiveLock() API

### PR DESCRIPTION
**Pull request by vrothberg**
Add a RecursiveLock() API to allow for recursive acquisitions of a
writer lock within the same process space.  This is yet another
requirement for the copy-detection mechanism in containers/image where
multiple goroutines can be pulling the same blob.  Having a recursive
lock avoids a complex synchronization mechanism as the commit order is
determinted by the corresponding index in the image.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@nalind @rhatdan @giuseppe PTAL

I need it [here](https://github.com/containers/image/blob/7531b7b2051bbd344ad65324e5a5a363e43c0d8a/storage/storage_image.go#L505).

**Comment by TomSweeneyRedHat** [comment]
LGTM, but definitely want some other head nods.

File: lockfile_unix.go [hunk 1]
```diff
@@ -161,7 +180,7 @@ func (l *lockfile) Unlock() {
 		// Close the file descriptor on the last unlock.
```
Comment (nalind) [inline lockfile_unix.go]: A `unix.FcntlFlock()` call to clear the lock should be unnecessary if we're going to close the descriptor right after.

Comment (vrothberg) [reply lockfile_unix.go]: That's unrelated to this PR. Let's fix that in another one.

**Comment by vrothberg** [comment]
@nalind @rhatdan are we good to go?

**Comment by vrothberg** [comment]
@nalind @rhatdan ping

**Comment by rhatdan** [comment]
I am fine with this, but we need @nalind review.

**Comment by nalind** [comment]
Looks reasonable, but I'd like a test that ensures that mixing recursive and non-recursive write locks in a single process doesn't unexpectedly break.

**Comment by vrothberg** [comment]
Great idea. I will update the PR and re-ping once the new tests are green.

File: lockfile_test.go [hunk 2]
```diff
@@ -408,6 +498,63 @@ func TestLockfileMixedConcurrent(t *testing.T) {
 	}
 }
 
+func TestLockfileMixedConcurrentRecursiveWriters(t *testing.T) {
```
Comment (vrothberg) [inline lockfile_test.go]: Here's the new test.

**Comment by rhatdan** [comment]
@nalind Can we merge this so we can create a new version?

**Comment by nalind** [comment]
LGTM

**Pull request merged**